strscan.gemspec

```diff
@@ -24,7 +24,7 @@ Gem::Specification.new do |s|
     s.files = %w{lib/strscan.jar lib/strscan.rb}
     s.platform = "java"
   else
-    s.files = %w{ext/strscan/extconf.rb ext/strscan/strscan.c}
+    s.files = %w{ext/strscan/extconf.rb ext/strscan/strscan.c lib/strscan.rb}
```

Comment:
I don't know why we need this.
If we don't install `lib/strscan.rb`, `strscan.rb` in TruffleRuby will be loaded. Because there are no files that match `require "strscan"` in the installed gem.

Reply:
One practical reason is otherwise `bundle exec rake test` can't work without this (because the file exists in the repo).
If the tests were actually done on an installed gem (would be good, but maybe as an extra CI step) that wouldn't be an issue, but that's uncommon and arguably impractical in development.

I think it is good to have `thisgem/lib/strscan.rb` in $LOADED_FEATURES, as a way to show the gem was loaded (and for TruffleRuby then delegated to stdlib).
In fact just looking at $LOADED_FEATURES shows what happened, which I find a very good thing.

Another reason is I feel is it's extremely confusing to only include some `lib/*.rb` on some gem platforms and not on others, nobody expects that and I'd think it also wouldn't work e.g. with `gem 'strscan', github: 'ruby/strscan'` or `gem 'strscan', path: '...'` in a Gemfile.

Reply:
I added a CI step to run tests on the installed gem, that seems a really good addition because it will be as close as we can to an actual usage of the gem.
That works fine except for 2.4 which doesn't have strscan as a default gem, so just skipped on 2.4.
(we can't use `bundle exec ruby run-test.rb` for 2.4 as that would add $PWD/lib to $LOAD_PATH and not load the gem, due to I think the `gemspec` in `Gemfile`)

Reply:
> One practical reason is otherwise `bundle exec rake test` can't work without this (because the file exists in the repo, regardless of the current gem platform).

One precision, for  `bundle exec rake test` we need to changes to `lib/strscan.rb` (as in this PR).
That itself doesn't force us to add to `lib/strscan.rb` so `s.files`.
But I find it very dangerous and brittle to have such a large difference between testing the gem locally with  `bundle exec rake test` and an actual real world usage of the gem. `lib/strscan.rb` should exist in both or neither, otherwise we're testing completely different things, and the other points I mentioned above apply as well.

Reply:
> If the tests were actually done only on an installed gem (would be good, but maybe as an extra CI step) that wouldn't be an issue, but that's uncommon and arguably impractical in development.

csv gem does this https://github.com/ruby/csv/blob/master/.github/workflows/test.yml#L74 because it can detect packaging problem. And it's useful to detect packaging problem before we release a gem.

> I think it is good to have `thisgem/lib/strscan.rb` in $LOADED_FEATURES, as a way to show the gem was loaded (and for TruffleRuby then delegated to stdlib).

Why is the information needed? It will just confuse users: "I want a new feature introduced by X.Y.Z but can't use the feature even when strscan X.Y.Z is loaded." Because TruffleRuby's bulit-in strscan may not be implemented the new feature yet. If `thisgem/lib/strscan.rb` isn't in `$LOADED_FEATURES`, users can think that "I can't use the new feature because strscan X.Y.Z isn't used".

> Another reason is I feel is it's extremely confusing to only include some `lib/*.rb` on some gem platforms and not on others, I think nobody expects that and I'd think it also wouldn't work e.g. with `gem 'strscan', github: 'ruby/strscan'` or `gem 'strscan', path: '...'` in a Gemfile.

I don't think so. (BTW, I feel that it's extremely confusing `gem "strscan", "= X.Y.Z"` is just ignored on TruffleRuby.)
I think that nobody doesn't care about it.

`gem 'strscan', path: '...'` doesn't work only on TruffleRuby. But I don't think it's a problem because TruffleRuby users don't need to use it. TruffleRuby always uses built-in strscan.

How about moving `lib/strscan.rb` to `lib/jruby/strscan.rb` and using `strscan.rb` only with JRuby? https://github.com/ruby/strscan/commit/0ab00f6fc44fbe22c578d92e50679d5aea9d2055

Reply:
> csv gem does this https://github.com/ruby/csv/blob/master/.github/workflows/test.yml#L74 because it can detect packaging problem. And it's useful to detect packaging problem before we release a gem.

That looks good, it's very similar to what I did but has a stronger guarantee the `lib/` isn't accidentally on `$LOAD_PATH` for that step.

> Why is the information needed? [...] users can think that "I can't use the new feature because strscan X.Y.Z isn't used".

Because users/developers might want to ensure the strscan gem is loaded/activated (e.g. I and other TruffleRuby devs certainly want to see that, it's very helpful for debugging what gets loaded, and e.g., the `puts $".grep(/\/strscan\./)` output would be unclear without this), otherwise they might think it's a bug of Bundler or of their usage (but it's not). The fact `strscan-1.x.y/lib/strscan.rb` is loaded, as well as `truffleruby/lib/truffle/strscan.rb` and not `strscan.so` should be pretty clear that the used code is from the truffleruby stdlib. I think this is by far the clearest result for everyone.
Not having a single file from the gem in $LOADED_FEATURES sounds like the gem was not used/activated at all, even though that's not true.
We can't see the gem in $LOAD_PATH (except by debugging in `lib/strscan.rb`), even though it was added to $LOAD_PATH, but we need to remove from $LOAD_PATH to actually load the stdlib. So at least we should preserve the trace that the gem was used/activated in $LOADED_FEATURES.

> BTW, I feel that it's extremely confusing gem "strscan", "= X.Y.Z" is just ignored on TruffleRuby.

It's not ignored with the lib/strscan.rb, and we'll see strscan-X.Y.Z is in $", that's part of the point :)
The behavior might not be completely the same, that's anyway true for any version due to a different implementation of StringScanner (and it's also true for the Java extension in this gem, unless you also modify it accordingly for every change).
I think users can quickly figure out if e.g. they use a new method and it works on CRuby but not on TruffleRuby it must TruffleRuby's fault. And even in case they file an issue here I'll handle it.

> How about moving `lib/strscan.rb` to `lib/jruby/strscan.rb` and using `strscan.rb` only with JRuby? [a2b7f6f](https://github.com/ruby/strscan/commit/a2b7f6f64a87069e223557d49619dd1b7dde73d0)

I dislike that, it's more magic, more difference between local development and the packaged gem ([this](https://github.com/ruby/strscan/commit/a2b7f6f64a87069e223557d49619dd1b7dde73d0#diff-188ab539bda7d5de390e7146eec51a5644a193d91e445103f86f4735eaaf0b9eR8) feels hacky to me), and it's brittle. For instance that would break if at some point TruffleRuby (or some other Ruby implementation) decides to implement some of StringScanner in a C extension in stdlib, because then the "wrong" strscan.so could get loaded by `require "strscan.so"` from truffleruby's strscan.rb (`require_relative` could help, but necessarily easy as `.so` are not always siblings of `.rb`).
Or if TruffleRuby (or some other Ruby implementation) implements all of StringScanner in some `strscan.so` or `strscan.someextensionext` then it would also break.
Or if there is at some point some Ruby file like `lib/strscan/foo.rb` in this gem, then it could be loaded from the gem instead of truffleruby stdlib, which is bad unless considered carefully whether that's safe.

And there is a performance advantage of having a lib/strscan.rb on CRuby too, it will avoid scanning the $LOAD_PATH for every `require "strscan"` after the first one: https://bugs.ruby-lang.org/issues/10902. That seems a significant advantage.

Ultimately if this is the approach you decide to take, then go ahead and I'll adapt the PR. However, if that approach is used I'm pretty sure I (and likely others too) will waste time to debug the magic of why there is no trace of the gem loaded in $", why local dev is so different than the real gem, etc. I think always having `lib/strscan.rb` is respecting the simplicity principle of programming.

Reply:
> Not having a single file from the gem in $LOADED_FEATURES sounds like the gem was not used/activated at all, even though that's not true.

We should use `Gem.loaded_specs` or something, or see `Gemfile.lock` instead of using `$LOADED_FEATURES` to confirm whether strscan gem is activated or not. Because they are the master information.

> We can't see the gem in $LOAD_PATH (except by debugging in `lib/strscan.rb`), even though it was added to $LOAD_PATH, but we need to remove from $LOAD_PATH to actually load the stdlib.

Generally, we should not modify `$LOAD_PATH` in a library to keep simple and easy to debug. (It's more magic. :-)

> The behavior might not be completely the same, that's anyway true for any version due to a different implementation of StringScanner (and it's also true for the Java extension in this gem, unless you also modify it accordingly for every change).

We'll keep synchronizing the C implementation and Java implementation for new behavior. I think that it's one of reasons that Java implementation joins to this repository.

> I dislike that, it's more magic, more difference between local development and the packaged gem ([this](https://github.com/ruby/strscan/commit/a2b7f6f64a87069e223557d49619dd1b7dde73d0#diff-188ab539bda7d5de390e7146eec51a5644a193d91e445103f86f4735eaaf0b9eR8) feels hacky to me), and it's brittle.

Sorry. It's my fault. I should have write it as https://github.com/ruby/strscan/commit/390cebbd00a75f1226c088237c29524723318449#diff-188ab539bda7d5de390e7146eec51a5644a193d91e445103f86f4735eaaf0b9eR6 to align the configuration in `.gemspec`.

> For instance that would break if at some point TruffleRuby (or some other Ruby implementation) decides to implement some of StringScanner in a C extension in stdlib, because then the "wrong" strscan.so could get loaded by `require "strscan.so"` from truffleruby's strscan.rb (`require_relative` could help, but necessarily easy as `.so` are not always siblings of `.rb`).

I can't understand this situation. Who does generate the "wrong" `strscan.so`? I think that this gem doesn't generate `strscan.so` with TruffleRuby with your `extconf.rb` change.

> Or if TruffleRuby (or some other Ruby implementation) implements all of StringScanner in some `strscan.so` or `strscan.someextensionext` then it would also break.

ditto.

> Or if there is at some point some Ruby file like `lib/strscan/foo.rb` in this gem, then it could be loaded from the gem instead of truffleruby stdlib, which is bad unless considered carefully whether that's safe.

CI will (should) detect it as a failure.

> And there is a performance advantage of having a lib/strscan.rb on CRuby too, it will avoid scanning the $LOAD_PATH for every `require "strscan"` after the first one: https://bugs.ruby-lang.org/issues/10902. That seems a significant advantage.

It seems that it's fixed in all maintained CRuby. So it doesn't affect to this case.

> Ultimately if this is the approach you decide to take, then go ahead and I'll adapt the PR. However, if that approach is used I'm pretty sure I (and likely others too) will waste time to debug the magic of why there is no trace of the gem loaded in $", why local dev is so different than the real gem, etc. I think always having `lib/strscan.rb` is respecting the simplicity principle of programming.

Using `$"` is a wrong approach to confirm loaded gems as I said above. And local dev isn't difference with the real gem. Both of them doesn't have `lib/strscan.rb` with `lib/jruby/strscan.rb` approach.

Reply:
> I can't understand this situation. Who does generate the "wrong" `strscan.so`? I think that this gem doesn't generate `strscan.so` with TruffleRuby with your `extconf.rb` change.

You're right, since this gem won't generate strscan.so on truffleruby (with this PR) that should no longer be an issue.

> CI will (should) detect it as a failure.

That's a bit hopeful, because it might appear to work but some part might not, but it's probably unlikely to have a `lib/strscan/foo.rb`.

> Both of them doesn't have `lib/strscan.rb` with `lib/jruby/strscan.rb` approach.

Alright, at least that's consistent between the gem and local dev. I'll cherry-pick your commit here and adapt the PR for it.

Reply:
@kou All done now, could you re-review?